subsystems/visionsubsystem.py

```diff
@@ -101,10 +109,143 @@ def periodic(self) -> None:
+        self.camToTargAngle = Rotation2d(
+            hypot(self.camToTargYaw.radians(), self.camToTargPitch.radians())
+        )
+
+        self.targToCam = Transform3d(targetPose, cameraPose)
+        self.targToCamYaw = Rotation2d(self.targToCam.X(), self.targToCam.Y())
+        self.targToCamPitch = Rotation2d(self.camToTargDistXY, -self.targToCam.Z())
+        self.targToCamAngle = Rotation2d(
+            hypot(self.targToCamYaw.radians(), self.targToCamPitch.radians())
+        )
+
+
+class SimCamera:
+    def __init__(
+        self,
+        name: str,
+        location: Transform3d,
+        horizFOV: float,
+        vertFOV: float,
+        key: str,
+    ) -> None:
+        self.name = name
+        self.location = location
+        self.horizFOV = horizFOV
+        self.vertFOV = vertFOV
+        self.key = key
+
+    def canSeeTarget(self, botPose: Pose3d, targetPose: Pose3d):
+        cameraPose = botPose + self.location
+        rel = CameraTargetRelation(cameraPose, targetPose)
+        return (
+            abs(rel.camToTargYaw.degrees()) < self.horizFOV / 2
+            and abs(rel.camToTargPitch.degrees()) < self.vertFOV / 2
+            and abs(rel.targToCamAngle.degrees()) < 90
+        )
+
+
+class VisionSubsystemSim(Subsystem):
+    def __init__(self) -> None:
+        Subsystem.__init__(self)
+        self.setName(__class__.__name__)
+
+        self.cameras = [
+            SimCamera(
+                name,
+                location,
+                constants.kCameraFOVHorizontal,
+                constants.kCameraFOVVertical,
+                key,
+            )
+            for name, location, key in zip(
+                constants.kPhotonvisionCameraArray,
+                constants.kCameraTransformsArray,
+                constants.kPhotonvisionKeyArray,
+            )
+        ]
+        self.poseList = []
+
+        self.rng = RNG(constants.kSimulationVariation, 500)
+
+    def periodic(self) -> None:
+        simPose = Pose2d(
+            *SmartDashboard.getNumberArray(constants.kSimRobotPoseArrayKey, [0, 0, 0])
         )
-        SmartDashboard.putNumberArray(camera.key, cameraPose)
+        simPose3d = pose3dFrom2d(simPose)
+
+        for camera in self.cameras:
+            seeTag = False
+            botPose = Pose3d()
+            for _id, apriltag in constants.kApriltagPositionDict.items():
+                if camera.canSeeTarget(simPose3d, apriltag):
+                    seeTag = True
+                    botPose = Pose3d(
+                        simPose3d.X() + self.rng.getNormalRandom(),
+                        simPose3d.Y() + self.rng.getNormalRandom(),
+                        simPose3d.Z() + self.rng.getNormalRandom(),
+                        simPose3d.rotation(),
+                    )
+
+            rel = CameraTargetRelation(simPose3d + camera.location, botPose)
+            VisionSubsystemReal.updateAdvantagescopePose(
+                botPose, camera.key, simPose3d, rel.camToTarg
```

Comment:
first argument should be the camera's pose

Reply:
![image](https://github.com/1757WestwoodRobotics/2024-Crescendo/assets/23267389/91e882c6-be92-471b-8345-4ea00abe0357)
axis on the ground comes from botPose